File: Problem-94-Bin_tree_inorder_traversal/optimized_sol.cpp
class Solution {
public:
    vector<int> inorderTraversal(TreeNode* root) {
        vector<int> output;
        TreeNode* curr = root;

        while (curr != nullptr) {
            if (curr->left == nullptr) {
                output.push_back(curr->val);
                curr = curr->right;
            } 
            else {
                TreeNode* pre = curr->left;
                while (pre->right && pre->right != curr) {
                    pre = pre->right;
                }

                if (pre->right == nullptr) {
                    pre->right = curr;  // make a thread
                    curr = curr->left;
                } else {
                    pre->right = nullptr;  // remove thread
                    output.push_back(curr->val);
                    curr = curr->right;
                }
            }
        }

        return output;
    }
};



// Time Complexity: O(n)

// Space Complexity: O(1)

File: Problem-1732-Find the Highest Altitude/optimized_sol.py
class Solution:
    def largestAltitude(self, gain: List[int]) -> int:
        current = 0
        maximum = 0

        for g in gain:
            current += g
            maximum = max(maximum, current)

        return maximum
    
# Time Complexity: O(n)
# Space Complexity: O(1)
